src/breakpoints.jl

```diff
@@ -0,0 +1,186 @@
+    set_breakpoint_active!(frame.code, active, pc)
+
+function get_breakpoint_active(framecode::JuliaFrameCode, pc)
+    stmtidx = convert(Int, pc)
+    return framecode.breakpoints[stmtidx][1]
+end
+get_breakpoint_active(frame::JuliaStackFrame, pc) =
+    get_breakpoint_active(frame.code, pc)
+
+function toggle_breakpoint_active!(framecode::JuliaFrameCode, pc)
+    stmtidx = convert(Int, pc)
+    active, condition = framecode.breakpoints[stmtidx]
+    framecode.breakpoints[stmtidx] = (!active, condition)
+    !active
+end
+toggle_breakpoint_active!(frame::JuliaStackFrame, active, pc) =
+    toggle_breakpoint_active!(frame.code, active, pc)
+
+enable(bp::Breakpoint)  = set_breakpoint_active!(bp.framecode, true, bp.stmtidx)
+disable(bp::Breakpoint) = set_breakpoint_active!(bp.framecode, false, bp.stmtidx)
+function remove(bp::Breakpoint)
+    bp.framecode.breakpoints[bp.stmtidx] = (false, falsecondition)
+    delete!(_breakpoints, bp)
+    return nothing
+end
+
+enable() = for bp in _breakpoints enable(bp) end
+disable() = for bp in _breakpoints disable(bp) end
+function remove()
+    for bp in _breakpoints
+        bp.framecode.breakpoints[bp.stmtidx] = (false, falsecondition)
+    end
+    empty!(_breakpoints)
+    return nothing
+end
+
+"""
+    breakpoint(f, sig)
+    breakpoint(f, sig, condition)
+    breakpoint(...; enter_generated=false)
+
+Add a breakpoint upon entry to `f` with the specified argument types `sig`.
+The first will break unconditionally, the second only if `condition` evaluates to `true`.
+`condition` should be written in terms of the arguments and local variables of `f`.
+
+# Example
+```julia
+function radius2(x, y)
+    return x^2 + y^2
+end
+
+breakpoint(radius2, Tuple{Int,Int}, :(y > x))
+```
+"""
+function breakpoint(f, sig::Type, condition::Union{Bool,Expr}=true; enter_generated=false)
+    method = which(f, sig)
+    framecode, _ = prepare_framecode(method, sig; enter_generated=enter_generated)
+    breakpoint!(framecode, 1, condition)
+end
+
+"""
+    breakpoint(method::Method)
+    breakpoint(method::Method, condition::Expr)
+
+Add a breakpoint upon entry to `method`. The first will break unconditionally, the second
+only if `condition` evaluates to `true`. `condition` should be written in terms of the
+arguments and local variables of `method`.
+"""
+function breakpoint(method::Method, condition::Union{Bool,Expr}=true)
+    # FIXME: this duplicates stuff in prepare_call
+    framecode = get(framedict, method, nothing)
+    if framecode === nothing
+        code = get_source(method)
+        framecode = JuliaFrameCode(method, code; generator=false)
+        framedict[method] = framecode
+    end
+    breakpoint!(framecode, 1, condition)
+end
+
+"""
+    breakpoint(f)
+    breakpoint(f, condition)
+
+Break-on-entry to all methods of `f`.
+"""
+function breakpoint(f, condition::Union{Bool,Expr}=true)
+    for method in methods(f)
+        breakpoint(method, condition)
```

Comment:
Im a bit worried about the number of breakpoints this might add in some cases. From a users point of view, if you add a breakpoint to `push!` you likely consider that one breakpoint that you handle and not `n` individual. 

Alternatively, it is up to the frontend to collapse all these breakpoints into one.

Reply:
I guess my question is if there should be "higher" breakpoint structs like

```
struct FunctionBreakpoint
    f::Function
    bps::Vector{JuliaFrameCode, Int}
end
```

Reply:
Im fine with the answer being that this is the low level end of things and that frontends should just wrap things and provide higher level abstractions themselves

Reply:
What's wrong with many breakpoints? Just that you have to prepare so many framecodes?

```julia
julia> using JuliaInterpreter

julia> foo(x) = x
foo (generic function with 1 method)

julia> breakpoint(foo)  # compile necessary methods

julia> @time breakpoint(push!)
  0.045100 seconds (15.96 k allocations: 977.291 KiB)

julia> length(Breakpoints._breakpoints)
22
```

Reply:
> From a users point of view, if you add a breakpoint to `push!` you likely consider that one breakpoint that you handle and not `n` individual.
> Alternatively, it is up to the frontend to collapse all these breakpoints into one.

So a user would likely want to be able to (something like)

```
bp = breakpoint(push!)
disable(bp)
```

in other words, handle the breakpoint on `push!` as it was a single breakpoint. Anyway, it is easy to workaround from a frontends pov so dont have to change anything here.

Reply:
Regardless of whether it's one meta-breakpoint or a vector of normal breakpoints, it would be useful to them here, instead of `nothing`.

Reply:
> it would be useful to them here

Did you accidentally a word here?

Reply:
Prob. "return them here"